### PR TITLE
Enhance Carousel Scrolling and Interaction in AttachmentCarousel

### PR DESCRIPTION
**Pull request by kidroca**
### Details

Key changes include:

- **Pan Gesture Integration:** Implemented a new gesture handler to allow smoother and more responsive scrolling through carousel items. This update leverages the `react-native-gesture-handler` for a fluid user experience.
  
- **Layout Optimization:** Adjusted the layout calculations to ensure attachments render correctly across different device sizes. This change utilises `useWindowDimensions` for dynamic layout updates based on screen size changes.

These updates address issues related to carousel bugs, noted in several related tickets (e.g., #23546, #39833, #22318, #21177, and #31166).

### Fixed Issues

$ https://github.com/Expensify/App/issues/23546
PROPOSAL: https://github.com/Expensify/App/issues/23546#issuecomment-2030160934

#### Verify Expected Behavior for Touch Screen Devices (iOS Safari, Android Chrome)

1. Open a chat containing multiple attachments.
2. Tap on any attachment to open the attachment carousel.
3. Swipe left or right to navigate through the attachments.
   - Ensure that swiping scrolls through the attachments one at a time and each attachment snaps into view centrally.

##### Bug #23546: Incorrect Image Displayed in HT Report
1. Open a chat with multiple attachments.
2. Open and close several attachments randomly.
3. Confirm that the attachment carousel always opens with the attachment that was selected.

##### Bug #22318: Attachment Skipping When Using Navigation Arrows
1. Open a chat with multiple attachments.
2. Select an attachment to open the attachment carousel.
3. Use the on-screen arrow keys to navigate through the attachments.
   - Verify that pressing the arrows moves the view to the next or previous attachment one at a time, without skipping.

##### Bug #31166: Inability to Open Protected PDFs in Chat View for Old Accounts
1. Open a chat containing password-protected PDF attachments.
2. Select a password-protected PDF to attempt opening it.
3. Enter the correct password in the provided field.
4. Tap on the eye icon within the password field to toggle visibility (ensure this does not trigger any bugs).
5. Click "Confirm" to proceed.
6. Verify that the PDF opens and can be viewed without issues.

### Offline tests

N/A

### QA Steps

Same as the `Tests` section above, same behavior is expected for staging and prod

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/12156624/fa28643c-36d1-4c5e-980f-0a65c2c8f6f9

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/12156624/e8950198-c3a6-4802-8385-5168eced9d29

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/12156624/85447058-0c4f-4673-83f2-bd36c4b859dd

https://github.com/Expensify/App/assets/12156624/7a726542-f3fa-4273-bdd9-e3a23999a07c

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/12156624/7d50e454-ff05-428b-a0c5-000cac70bdf1

https://github.com/Expensify/App/assets/12156624/73598345-c321-4dbc-94c5-5880d1ba3a8f

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/12156624/9ef82e3d-57f5-42a3-b8d7-d6889e9a086c

https://github.com/Expensify/App/assets/12156624/dfccdddf-2c0e-4e98-88da-b7a1659e1705

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/12156624/104880d4-0ae9-450b-bd32-a02d16e7b91e

</details>

### CODE REVIEW

File: src/components/Attachments/AttachmentCarousel/AttachmentCarouselCellRenderer.tsx
[the remark this answers is not in the document]
Comment (kidroca) [reply src/components/Attachments/AttachmentCarousel/AttachmentCarouselCellRenderer.tsx:1]: We need accurate cell width to make the pan gesture responder work, so we moved the calculation to happen on the list level. This makes the custom cell renderer component redundant and have the benefit of not doing the same calculation per each rendered cell

File: src/components/Attachments/AttachmentCarousel/index.tsx [hunk 1]
```diff
@@ -29,16 +31,23 @@ const viewabilityConfig = {
     itemVisiblePercentThreshold: 95,
 };
 
+const MIN_FLING_VELOCITY = 500;
```
[the remark this answers is not in the document]
Comment (kidroca) [reply src/components/Attachments/AttachmentCarousel/index.tsx:34]: This custom value is used to gauge the speed of a user's swipe:
- A quick lateral swipe advances to the previous or next image based on the direction.
- A slow drag that releases after moving the image more than 50% of its width will also transition to the previous or next image.
- A slow drag that moves a minimal distance and is then abandoned (e.g., less than 50%) will cause the image to snap back to the center.

I determined this value through testing typical swiping motions intended to fling to the next image. However, I am open to suggestions for an optimized value. Given that Software Mansion is the creator of the gesture handler used here and is involved in this project, maybe we can give them a shout?

**Comment by kidroca** [comment]
I'll investigate the issue and try to find a fix. However, since the bugs don't appear to be critical or common, it might be more efficient to address them in a follow-up PR. Can you confirm what would be the priority?

**Comment by CortneyOfstad** [comment]
@parasharrajat any update on @kidroca's comment [here](https://github.com/Expensify/App/pull/39930#issuecomment-2091121753)?

**Comment by kidroca** [comment]
> Can we solve [#39930 (comment)](https://github.com/Expensify/App/pull/39930#issuecomment-2087004800) @kidroca for now? This looks like a thing that will become a issue for Desktop users.

Synced with `main` and pushed an update to fix issues around resizing: 3a1e687de0db447a7be7e2a74a04ad1e6e1b48e1

**Comment by kidroca** [comment]
> I was able to reproduce this [#39930 (comment)](https://github.com/Expensify/App/pull/39930#issuecomment-2088440251) again so it seems that it will be a real issue for users.

I couldn't reproduce this bug. I was able to tap, hold, and prevent the transition to the next page after pressing the next arrow, but it didn't result in any unusual behaviour.

Could you describe the problem you're experiencing after pressing the next button and then trying to drag? In the video, it seems like the pages are no longer aligning correctly, but I can't tell if that's the case or if you're still dragging. Do you suggest we should disable manual scrolling while the transition to the next page is running?

**Review by parasharrajat — APPROVED** [review]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

:ribbon: :eyes: :ribbon:  C+ reviewed

**Review by deetergp — APPROVED** [review]
Looks good and tests well. Not merging because we are in a merge freeze till at least next week.

**Comment by deetergp** [comment]
Okay, we have thawed out from our freeze, so I am merging this now.

**Pull request merged**